Add JSON AST dumps

sgizler opened the pull request:
Verilator's AST can currently be dumped into custom `.tree` representation for debug purposes. This PR adds a mostly equivalent JSON representation which is easier to process programmatically.

## Example use cases

### Tree diffing
An example of how this is useful is the [`astsee`](https://github.com/antmicro/astsee) tool which can pretty-print and diff these JSON dumps.

![diff](https://github.com/antmicro/verilator-1/assets/9216518/4a07ef4f-8b9a-47d7-9c72-181ae6125af5)

### Tree navigation
An example HTML representation for easy navigation of the AST, generated by `astsee`:

[example.zip](https://github.com/antmicro/verilator-1/files/13439928/example.zip)

### GDB commands
`astsee` can also be used from within GDB. Useful for diffing the AST at runtime, even within a single Verilator stage.

![gdb](https://github.com/antmicro/verilator-1/assets/9216518/f1821899-2e09-40c2-b898-dc31300141b3)

wsnyder commented:
What is the use case?

Why can the existing XML format not be used instead?

kbieganski commented:
> What is the use case?

There are examples in the PR description.

> Why can the existing XML format not be used instead?

It affects verilation (disables stages, etc). This is an alternative to `--dump-tree`.

wsnyder commented:
Sorry, I should have been more explicit, why can the existing XML format not get an option added to dump it at every stage, instead of adding a new JSON format that is dumped at every stage?

kbieganski commented:
* XML is pretty rare these days. JSON has a healthier ecosystem, at least in the open source space (e.g. tools like `jq`, various viewers, converters...)
* One of the XML dump visitors manipulates `user1` (so it cannot be used inside a stage)
* A lot of info is missing from the XML, and we don't know if it should stay that way or not. Just seems like a different use case
* The XML dump has different names for some node fields
* The JSON dump implementation is very simple compared to the multiple XML visitors

wsnyder commented:
XML and JSON are both serialization methods, and we can "fundamentally" use either for the two use cases:
1. A file that is somewhat more stable than the AST, is stable run-to-run, but it is still coupled somewhat tightly to Verilator version.  (e.g. --xml-only presently)
2. A format that represents all debug info in a machine readable format and shows exact pointers. (e.g. your JSON proposal here).

But nothing prevents XML for output 2, or JSON for output 1 (e.g. a --json-only).

Some concerns:
* JSON is the serialization-flavor-of-the-present, just like XML used to be.  Not really relevant, just saying.
* Once we have a machine readable format people will have apps that use one or the other format, and we'll get requests to enhance one or the other, and so:
* I dislike maintaining multiple serialization formats.
  Can/should we add a --json-only, change the example scripts to read JSON, and mark XML for deprecation in 2025?
  The main difference between the --json-only's JSON should be we don't dump pointers and instead stable identifiers.  Perhaps we need a --json-abstract flag that does that disambiguation and makes all pointers into stable IDs.
* JSON is much larger dump-wise, and on large designs the space can be significant. For the debug tools you discussed, which are useful, they are all the more useful on larger designs, so we want something efficient.
  What can we do to shrink the output?
  Can we reduce whitespace, and have single-space indent per level?
  Do the downstream tools accept json.gz?

As to patch itself, please rebase to master so I can try it more easily.

kbieganski commented:
> 1. A file that is somewhat more stable than the AST, is stable run-to-run, but it is still coupled somewhat tightly to Verilator version.  (e.g. --xml-only presently)
> 
> 2. A format that represents all debug info in a machine readable format and shows exact pointers. (e.g. your JSON proposal here).

The first use case just seems like a completely different feature due to the stability requirement (I assume you meant stability of structure between releases, not just pointers).

> * JSON is the serialization-flavor-of-the-present, just like XML used to be.  Not really relevant, just saying.

I think it is relevant, as moving to a format that's going out of fashion would be a bad idea. However, JSON doesn't really have significant competition in the human-readable serialization space, for better or for worse. It's been around for about 20 years, and I expect it to still be around for at least another 20 years.

> * I dislike maintaining multiple serialization formats.

Completely understandable. But if `--xml-only`/`--json-only` needs to be more stable in terms of its definition, then we'll still have basically two serialization formats. Also, looking at `V3EmitXml.cpp` commit history, it doesn't seem like it's causing much maintenance overhead.

> Can/should we add a --json-only, change the example scripts to read JSON, and mark XML for deprecation in 2025?

I find JSON easier to work with, so I'm for it, but I've also never used `--emit-xml`.

> The main difference between the --json-only's JSON should be we don't dump pointers and instead stable identifiers.  Perhaps we need a --json-abstract flag that does that disambiguation and makes all pointers into stable IDs.

Stable identifiers would probably be a good default for either case. Currently there's an option to just filter out pointers.

> * JSON is much larger dump-wise, and on large designs the space can be significant. For the debug tools you discussed, which are useful, they are all the more useful on larger designs, so we want something efficient.

Larger than `--dump-tree`, right? XML is slightly larger than JSON due to the end tags, but not significantly. `--dump-tree-json` is a little over two times bigger than `--dump-tree`. So yes, not very useful for debugging large designs at this point, but neither are the other options.

> What can we do to shrink the output?
> Can we reduce whitespace, and have single-space indent per level?

Sure, whitespace is irrelevant in JSON.

> Do the downstream tools accept json.gz?

Not that I know of, but you can easily pipe it through `gunzip`. I do agree that compressed dump output is worth exploring.

gezalore commented:
> Stable identifiers would probably be a good default for either case. Currently there's an option to just filter out pointers.

FWIW see --dump-tree-addrids for a noddy implementation of stable identifiers in existing dumps

wsnyder commented:
Note -xml-only does not make version-stability guarantees.

Ok, I accept this concept, I'd like the pull to:
* Add json-only as conceptual replacement of --xml-only (or as a separate pre-pull), with a similar or superset feature set
* Replace existing examples/xml_py to read JSON.
* Indicate XML is deprecated (no sooner than Jan 2025 assuming this patch is merged by Jan 24)
* Have JSON whitespace with one space indent per-level, and one line per node (I use "grep" a lot to filter files)
* Honor --dump-tree-addrids; have this on by default at least with --json-only  (or always on by default for JSON?)

wsnyder commented:
@wallento @Risto97 @jameshanlon @pieter3d, who all used used XML in the past, for any comments as to plans to replace XML with JSON.

pieter3d commented:
Sounds like a good idea.  I don't have any active flows or projects that rely on xml.

jameshanlon commented:
The proposal is fine by me since I am no longer actively maintaining netlist-paths. I've developed a similar project using slang, which incidentally has a json AST serialiser.

sgizler commented on src/V3Options.cpp:
```diff
@@ -881,11 +881,19 @@ void V3Options::notify() VL_MT_DISABLED {
 
+    // --dump-json-tree aliases --dump-tree-json, as it's an easy mistake
+    if (m_dumpLevel.count("json-tree")) m_dumpLevel["tree-json"] = m_dumpLevel["json-tree"];
+
```
Alias removed.


Sidenote: I initially added as kind of workaround of forgiving parsing of `--dump*` options (you can pass `--dump-anything` and hear no complain about missing opt from verilator)

sgizler commented on src/V3AstNodes.cpp:
```diff
@@ -27,8 +27,47 @@
+// strings with a bijection to aid human readability. Observe that this might
+// not actually be a unique identifier as the address can get reused after a
+// node has been freed.
+static std::string nodeAddr(const AstNode* nodep) {
```
As of now I made it a `public static` member of `AstNode` as it is used in `dumpJsonPtr()` "global" func

sgizler commented on test_regress/t/t_dump_json.pl:
Fixed by adding `--json-fileids` (that works in similar fashion to `--dump-tree-addrids`) that allowed me to get rid of `jq` postprocessing. See https://github.com/verilator/verilator/pull/4715/commits/15a756dd8947354b76e05e15969e39fb685ad6ac

sgizler commented on src/V3Options.cpp:
```diff
@@ -1173,6 +1181,7 @@ void V3Options::parseOptsList(FileLine* fl, const string& optdir, int argc,
     DECL_OPTION("-dumpi-", CbPartialMatchVal, [this](const char* optp, const char* valp) {
         m_dumpLevel[optp] = std::atoi(valp);
     });
+    DECL_OPTION("-no-json-addrs", Set, &m_noJsonAddrs);
```
> I think we should default this as on. 

Since `--dump-tree-addrids` works, I think it is superior default to `--no-json-addr` (well, it even makes it kind of redundant).

sgizler replied:
I made `--dump-tree-addrids` a default.

wsnyder commented:
Please merge from master. Now with #4727, you should be able to automate the JSON dumping of pointers by making a dumpJsonGen function in astgen.  Other member dumps will need to remain in e.g. dumpJson.

wsnyder commented:
Note I'm working on another pull that will use this new function in astgen; I think you can use the identical function called with an appropriate Lambda for JSON dumping.

```
diff --git a/src/astgen b/src/astgen
index a51d14a3f..141c7d4d7 100755
--- a/src/astgen
+++ b/src/astgen
@@ -1020,6 +1020,17 @@ def write_ast_impl(filename):

             emitBlock("}}\n")

+            emitBlock(
+                "void Ast{t}::foreachLink(std::function<void(AstNode*, const char*)> f) const {{\n",
+                t=node.name)
+            if node.superClass.name != 'Node':
+                emitBlock("    Ast{base}::foreachLink(f);\n",
+                          base=node.superClass.name)
+            for ptr in node.ptrs:
+                # FIXME why compile error
+                emitBlock("    f(m_{name}, \"{name}\");", name=ptr['name'])
+            emitBlock("}}\n")
+

 def write_ast_macros(filename):
     with open_file(filename) as fh:
@@ -1066,6 +1077,7 @@ def write_ast_macros(filename):
             Ast{t}* addNext(Ast{t}* nodep) {{ return static_cast<Ast{t}*>(AstNode::addNext(this, nodep)); }}
             const char* brokenGen() const override;
             void cloneRelinkGen() override;
+            void foreachLink(std::function<void(AstNode*, const char*)> f) const override;
             ''',
                       t=node.name)



```

wsnyder reviewed (changes requested):
Great stuff. I didn't get fully through the patch set, but here's my first round of comments.

sgizler commented on bin/verilator:
```diff
@@ -334,10 +334,14 @@ detailed descriptions of these arguments.
     --dump-tree                 Enable dumping Ast .tree files
     --dump-tree-addrids         Use short identifiers instead of addresses
+    --dump-tree-json            Enable dumping Ast .tree.json files and .tree.meta.json file
+    --no-json-ids               Don't use short identifiers instead of adresses/paths in .tree.json
+    --no-json-edit-nums         Don't dump editNum in .tree.json files
```
You mean ordering?

wsnyder replied:
Yes, sort as if "json-ids", "json-edit-nums".  Note also flip them as "e" before "i" ;)

wsnyder commented on test_regress/t/t_dump_json.out:
```diff
@@ -0,0 +1,526 @@
+{"type":"NETLIST","name":"$root","addr":"(B)", "loc":"a,0:0,0:0","timeunit":"1ps","timeprecision":"1ps","typeTablep":"(C)","constPoolp":"(D)","dollarUnitPkgp":"UNLINKED","stdPackagep":"UNLINKED","evalp":"UNLINKED","evalNbap":"UNLINKED","dpiExportTriggerp":"UNLINKED","delaySchedulerp":"UNLINKED","nbaEventp":"UNLINKED","nbaEventTriggerp":"UNLINKED","topScopep":"UNLINKED",
```
Ok, never mind then.

wsnyder reviewed (commented):
It looks like test_regress/t/t_constraint_xml* were removed by accident.

wsnyder commented on src/V3Active.cpp:
```diff
@@ -648,5 +648,5 @@ class ActiveVisitor final : public VNVisitor {
     UINFO(2, __FUNCTION__ << ": " << endl);
     { ActiveVisitor{nodep}; }  // Destruct before checking
-    V3Global::dumpCheckGlobalTree("active", 0, dumpTreeLevel() >= 3);
+    V3Global::dumpCheckGlobalTree("active", 0, dumpTreeEitherLevel() >= 3);
```
Thanks for renaming this.  Can you please make a new pull request just with this rename across all of the appropriate functions, and the function declared to match dumpTreeLevel()?  This will greatly shrink #4715 patch size.  Thanks.

(Note one way to make this is just do the "git diff" to a file and manually edit to just have the patches that create this function.)

kbieganski commented on src/V3Options.h:
```diff
@@ -514,6 +519,8 @@ class V3Options final {
     bool xInitialEdge() const { return m_xInitialEdge; }
     bool xmlOnly() const { return m_xmlOnly; }
+    bool jsonOnly() const { return m_jsonOnly; }
+    bool xmlOrJsonOnly() const { return m_xmlOnly || m_jsonOnly; }
```
This is a bit different than what we usually mean by 'dump'. Not a fan of overloading the term. But I can't come up with a better one. Maybe `elabOnly`? That's a bit too vague.

sgizler replied:
>serializeOnly?
sounds nice

sgizler commented:
> It looks like test_regress/t/t_constraint_xml* were removed by accident.

They were superseded by test_regress/t/t_constraint_json_only*

sgizler commented:
Sqashed commits a bit and rebased on master. 

Opened separate PR #4831 that only adds --`json-only`. 
`--dump-tree-json` and gdb commands will be added by separate PR (probably this one) after #4831

sgizler commented:
> Opened separate PR https://github.com/verilator/verilator/pull/4831 that only adds --json-only.
--dump-tree-json and gdb commands will be added by separate PR (probably this one) after https://github.com/verilator/verilator/pull/4831

#4831 was merged.

I rebased this PR on changes introduced there, added test for gdb commands and installed `astsee` in CI

wsnyder commented:
This was merged in last version, if changes are needed please make a new pull, and delete the old branch (after copying any needed work out of it), thanks.

wsnyder commented:
This PR was referenced as the one in the Changelog, sounds like a mistake based on your comment, but given that I'd prefer a new pull as I can't backout last release ;).

kbieganski commented:
Opened a new PR: https://github.com/verilator/verilator/pull/5020
